### PR TITLE
IA-2292 500 when using a csv with semicolon in batch user creation + add projects and user roles

### PR DESCRIPTION
**Pull request by SimonDus**
Explain what problem this PR is resolving

Related JIRA tickets : IA-2292 and IA-2289

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are there enough tests

## Changes

Field Updates : 

Mandatory columns  have been updated to make it less error prone. 

New column " user_roles " and " projects " have been added.

Mandatory columns are : ["username", "password", "email", "first_name", "last_name", "orgunit", "profile_language", "dhis2_id", "permissions", "user_roles", "projects"]

I added a check of csv delimiter to handle files separated by both coma and semicolon. For some reason sometimes ne Sniffer doesnt work and result into a error. So the code check first with Sniffer, if it fails it try a " manual " check by parsing the file. 

Also, more tests.

## How to test

Explain how to test your PR.
Test with a CSV created with Excel.
Upload a CSV file with semicolon separator, it should work. Also test the new fields. Use one of the working templates to test.

### CODE REVIEW

**Review by hakifran — CHANGES_REQUESTED** [review]
While testing there is a bug 
Screenshot from 2023-07-05 11-14-21

File: iaso/api/bulk_create_users.py [hunk 1]
```diff
@@ -88,17 +88,24 @@ def create(self, request, *args, **kwargs):
...
-            i = 0
             csv_indexes = []
             file_instance = BulkCreateUserCsvFile.objects.create(
                 file=user_csv, created_by=request.user, account=request.user.iaso_profile.account
             )
             file_instance.save()
-            for row in reader:
+            for i, row in enumerate(reader):
                 org_units_list = []
                 if i > 0:
```
Comment (madewulf) [inline iaso/api/bulk_create_users.py:158]: Wondering if we should not have an except for all exceptions and raising then an error saying "there is a problem with your file" ?

**Comment by SimonDus** [comment]
@hakifran  I just did a test and I don't have the error. Can you describe me what you are doing ?

File: iaso/api/bulk_create_users.py (hunk 1 above)
Comment (SimonDus) [reply iaso/api/bulk_create_users.py:158]: I added more check to the csv files/Fields etc. It should be more secure now

**Comment by SimonDus** [comment]
I just found a bug that I have to fix.

**Review by hakifran — CHANGES_REQUESTED** [review]
@SimonDus , when a user role it's created there is a prefix which is added on the user role name. The prefix is the account id and an underscore like '1_manager'. So in the csv file if the prefix is not added it won't find the user role.Could you fix it?
Screenshot from 2023-08-01 14-29-42

Screenshot from 2023-08-01 14-30-21

**Review by hakifran — APPROVED** [review]
On my point of view it's good

**Review by mathvdh — APPROVED** [review]
Seems good to me !

**Comment by mathvdh** [comment]
Juste une question : pourquoi il m'affiche ce message bizarre tout en haut de cette page ? juste parce qu'il y  a un `;` dans le nom de la branche ?

Screenshot 2023-08-04 at 08 44 17

**Comment by SimonDus** [comment]
@mathvdh Oui c'est juste à cause du " ; " c'est pas bien grave.

**Pull request merged**